MapleServer2/PacketHandlers/Game/MasteryHandler.cs

```diff
@@ -79,19 +87,39 @@ private void HandleCraftItem(GameSession session, PacketReader packet)
                 return;
             }
 
+            if (recipe.RequireMastery > 0)
+            {
+                if (session.Player.Levels.MasteryExp.FirstOrDefault(x => x.Type == (MasteryType) recipe.MasteryType).CurrentExp < recipe.RequireMastery)
+                {
+                    session.Send(MasteryPacket.MasteryNotice((short) MasteryNotice.NotEnoughMastery));
+                    return;
+                }
+            }
+
+            if (recipe.RequireQuest.Count > 0)
+            {
+                foreach (int questId in recipe.RequireQuest)
+                {
+                    QuestStatus quest = session.Player.QuestList.FirstOrDefault(x => x.Basic.Id == questId);
+                    if (quest == null || !quest.Completed)
+                    {
+                        session.Send(MasteryPacket.MasteryNotice((short) MasteryNotice.RequiredQuestIsNotCompleted));
+                        return;
+                    }
+                }
+            }
+
             // does the play have enough mesos for this recipe?
-            if (!session.Player.Wallet.Meso.Modify(-recipe.GetMesosRequired()))
+            if (!session.Player.Wallet.Meso.Modify(-recipe.RequireMeso))
```

Comment:
should `RequireMeso` be instead `RequiredMeso`?

Reply:
sure, I just kept it line with what was currently there in the RecipeMetadata. We currently have all requirements as "RequireMastery","RequireQuest","RequireItem" etc. should we change them all to Required?

Reply:
Nah nevermind